``pyreverse``: add PlantUML output

DudeNr33 opened the pull request:
- [x] Add a ChangeLog entry describing what your PR does.
- [x] If it's a new feature, or an important bug fix, add a What's New entry in
      `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Type of Changes

<!-- Leave the corresponding lines for the applicable type of change: -->

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description
This MR adds output in PlantUML format to ``pyreverse``.
I wanted to make sure that the feature from #4551 also works properly with PlantUML, so I added a new class to the test files.
This made the list of changed files a bit longer as intended. 

Closes #4498

DudeNr33 commented on pylint/pyreverse/dot_printer.py:
```diff
@@ -103,31 +102,11 @@ def _build_label_for_node(
-                for arg, ann in annotations.items()
-            )
-
-            label += fr"{func.name}({args}){return_type}\l"
```
The same logic is used by ``PlantUmlPrinter``, so I extracted a method and put it inside the base ``Printer`` class.

DudeNr33 commented on pylint/pyreverse/main.py:
```diff
@@ -209,8 +207,7 @@ def run(self, args):
             sys.path.pop(0)
-        printer_class = VCGPrinter if self.config.output_format == "vcg" else DotPrinter
-        writer.DiagramWriter(self.config, printer_class).write(diadefs)
+        writer.DiagramWriter(self.config).write(diadefs)
```
I introduced a factory function to get the printer class for a given filetype. This supports the Open-Closed-Principle (only factory function needs to be updated if a new printer is added). Also, the ``DiagramWriter`` can now obtain the printer class itself.

DudeNr33 commented on pylint/pyreverse/printer.py:
```diff
@@ -84,6 +86,28 @@ def emit_edge(
+            f"{arg.name}: {ann}" if ann else f"{arg.name}"
+            for arg, ann in annotations.items()
+        ]
+
```
This method is used by both ``DotPrinter`` and ``PlantUmlPrinter``. We could probably add this feature for VCG as well if somebody needs it.

DudeNr33 commented on tests/data/suppliermodule_test.py:
```diff
@@ -10,3 +10,11 @@ def set_value(self, value):
+        self.my_int = optional_int
+
+    def do_it(self, new_int: int) -> int:
+        return self.my_int + new_int
```
This new class is for checking that #4551 works correctly with PlantUML output too.

DudeNr33 commented on tests/pyreverse/test_writer.py:
```diff
@@ -110,22 +114,22 @@ def _setup(project, config, writer):
-    )
-    assert expected == generated, f"{files}{diff}"
-    os.remove(generated_file)
+    _assert_files_are_equal(generated_file)
```
The logic for each test case is basically the same, but the used fixtures differ. To avoid duplicated code I extracted a helper function for it. 
The ``setup_xy`` are fixtures as they 
a) shall perform cleanup code after the test case finishes
b) test scope is comparing the output file with expectation, not the process of generating the files itself

Is there a way to specify the fixture to use in ``pytest.mark.parametrize``?

DudeNr33 commented on tests/pyreverse/test_writer.py:
```diff
@@ -136,4 +140,3 @@ def test_vcg_files(generated_file):
         line for line in unified_diff(expected.splitlines(), generated.splitlines())
     )
     assert expected == generated, f"{files}{diff}"
-    os.remove(generated_file)
```
Cleanup is performed in the fixture, this ``os.remove`` is not necessary.

Pierre-Sassoulas commented on pylint/pyreverse/printer_factory.py:
```diff
@@ -0,0 +1,17 @@
+from pylint.pyreverse.printer import Printer
+from pylint.pyreverse.vcg_printer import VCGPrinter
+
+filetype_to_printer = {"vcg": VCGPrinter, "puml": PlantUmlPrinter, "dot": DotPrinter}
```
I often use ``.plantuml`` personally, I guess adding that won't hurt.
```suggestion
filetype_to_printer = {
    "vcg": VCGPrinter,
    "puml": PlantUmlPrinter,
    "plantuml": PlantUmlPrinter,
    "dot": DotPrinter,
}
```

Pierre-Sassoulas commented on tests/pyreverse/test_printer_factory.py:
```diff
@@ -0,0 +1,28 @@
+    [
+        ("vcg", VCGPrinter),
+        ("dot", DotPrinter),
+        ("puml", PlantUmlPrinter),
```
```suggestion
        ("puml", PlantUmlPrinter),
        ("plantuml", PlantUmlPrinter),
```

Pierre-Sassoulas reviewed (approved):
This refactor + new plantuml writter look very good ! Thank you for being so persistent, for separating this huge feature into multiple merge request and for all the explanations you gave. It's definitely going to be one of the major feature of 2.10.

Pierre-Sassoulas commented:
Maybe we could try covering the 3 lines not covered in the new plantuml writter with a diagram that would trigger them?

DudeNr33 commented:
Thanks for all your reviews, too! 😁 
I added more tests (one of which actually caught a bug!).
But one line won't be covered until I have added the last part of the original MR: colored output. 
Fontcolor was already used for exceptions (which I added to the test data), but body color is not used yet.

Pierre-Sassoulas commented:
>  one line won't be covered until I have added the last part of the original MR: colored output.

Understood, let's merge this then :)
